src/uu/dd/src/datastructures.rs

```diff
@@ -0,0 +1,171 @@
+// This file is part of the uutils coreutils package.
+//
+// (c) Tyler Steele <tyler.steele@protonmail.com>
+//
+// For the full copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+use crate::conversion_tables::*;
+
+use std::error::Error;
+use std::time;
+
+pub struct ProgUpdate {
+    pub read_stat: ReadStat,
+    pub write_stat: WriteStat,
+    pub duration: time::Duration,
+}
+
+#[derive(Clone, Copy, Default)]
+pub struct ReadStat {
+    pub reads_complete: u64,
+    pub reads_partial: u64,
+    pub records_truncated: u32,
+}
+impl std::ops::AddAssign for ReadStat {
+    fn add_assign(&mut self, other: Self) {
+        *self = Self {
+            reads_complete: self.reads_complete + other.reads_complete,
+            reads_partial: self.reads_partial + other.reads_partial,
+            records_truncated: self.records_truncated + other.records_truncated,
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct WriteStat {
+    pub writes_complete: u64,
+    pub writes_partial: u64,
+    pub bytes_total: u128,
+}
+impl std::ops::AddAssign for WriteStat {
+    fn add_assign(&mut self, other: Self) {
+        *self = Self {
+            writes_complete: self.writes_complete + other.writes_complete,
+            writes_partial: self.writes_partial + other.writes_partial,
+            bytes_total: self.bytes_total + other.bytes_total,
+        }
+    }
+}
+
+type Cbs = usize;
+
+/// Stores all Conv Flags that apply to the input
+#[derive(Debug, Default, PartialEq)]
+pub struct IConvFlags {
+    pub ctable: Option<&'static ConversionTable>,
+    pub block: Option<Cbs>,
+    pub unblock: Option<Cbs>,
+    pub swab: bool,
+    pub sync: Option<u8>,
+    pub noerror: bool,
+}
+
+/// Stores all Conv Flags that apply to the output
+#[derive(Debug, Default, PartialEq)]
+pub struct OConvFlags {
+    pub sparse: bool,
+    pub excl: bool,
+    pub nocreat: bool,
+    pub notrunc: bool,
+    pub fdatasync: bool,
+    pub fsync: bool,
+}
+
+/// Stores all Flags that apply to the input
+#[derive(Debug, Default, PartialEq)]
+pub struct IFlags {
+    pub cio: bool,
+    pub direct: bool,
+    pub directory: bool,
+    pub dsync: bool,
+    pub sync: bool,
+    pub nocache: bool,
+    pub nonblock: bool,
+    pub noatime: bool,
```

Comment:
Please add the [unknown words](https://github.com/uutils/coreutils/runs/3131291167?check_suite_focus=true) to spell checker allow list, or choose different words that don't trigger the spell checker.

Reply:
I believe I got them all. We'll see next push :crossed_fingers: